[24.04] Osrfbuild migrate Linux agents recipe

Crola1702 opened the pull request:
# Description

This PR updates the cookbook to be able for using it on 24.04 linux machines

## Changes

- Upgrade to jdk-21-jre
- Upgrade to swarm plugin 3.49
- Use package `docker.io`, which is the same package used in [ros buildfarm agents](https://github.com/ros-infrastructure/cookbook-ros-buildfarm/blob/8d2a7280683118c5e624bd734ecb1a2c653240c1/recipes/agent.rb#L2)
- Remove python2 packages that don't exist in ubuntu 24.04
- Install `ubuntu-drivers-common` which is needed to provide `/usr/bin/gpu-manager`, which is also needed to run `lightdm`.
- Install `nvidia-container-toolkit` from stable debs repositories. Also, add a configuration step for it to allow docker to use nvidia runtime
- Change the location where `gdm3` is purged.
   - In this case, the purge instruction for gdm3 was doing nothing, because gdm3 isn't installed when chef runs that instruction. I think gdm3 was installed after somehow. Now that the location has changed, I can see the purge action take effect in the packer run.
- Restart `accounts-daemon` before lightdm. This way we prevent: `Error updating user /org/freedesktop/Accounts/User1000: GDBus.Error:org.freedesktop.DBus.Error.InvalidArgs: No such interface ?org.freedesktop.DisplayManager.AccountsService?`

Crola1702 committed:
Update apt packages

- Use docker.io instead of docker-ce
- Remove apt packages not present in noble

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Update java and swarm versions

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Use nvidia-container-toolkit

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Add drivers-common to provide /usr/bin/gpu-manager

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Specify BusID for nvidia GPU xorg config

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Delete commented code

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Purge gdm3 after lightdm is installed

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Ensure accounts-daemon is restarted before lightdm

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Remove unused code

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

j-rivero commented on files/xorg.conf.nvidia:
```diff
@@ -36,6 +36,7 @@ Section "Device"
     Identifier     "Device0"
     Driver         "nvidia"
     VendorName     "NVIDIA Corporation"
+    BusID          "PCI:0:30:0"
```
Did you find this change required to make the GPU to work? Making the BusID explicit make us less resilient to changes in the hardware.

nuclearsandwich replied:
The chef-iest way to do this would be to use a custom Ohai plugin https://docs.chef.io/ohai_custom/ to gather the information we need when we initially profile the host. A slightly less idiomatic way that might be more expedient given that we use chef-solo would be to use a Ruby block resource that pulls the information from e.g. `lspci` into `node.run_state`.

j-rivero committed:
Update CI to 24.04

Signed-off-by: Jose Luis Rivero <jrivero@honurobotics.com>

j-rivero commented:
Change the CI to 24.04. 

If you install kitchen in your local machine that can work using `kitchen list`, `kitchen test` or `kitchen converge`. The CI has been really useful in this not extremely active repositories to be sure that changes are not breaking current setup. Let's try to keep it working.

j-rivero reviewed (changes requested):
Ci is complaining about:

```
       [2025-06-06T12:00:39+00:00] FATAL: Chef::Exceptions::ResourceNotFound: resource cookbook_file[/etc/lightdm/xhost.sh] is configured to notify resource service[accounts-daemon] with action restart, but service[accounts-daemon] cannot be found in the resource collection. cookbook_file[/etc/lightdm/xhost.sh] is defined in /opt/kitchen/cache/cookbooks/osrf_jenkins_agent/recipes/agent_build_tools.rb:144:in `from_file'
```

Crola1702 commented:
> Ci is complaining about:

Hmm, do you know if dokken supports linux native systemd units such as `accountsservice` which provides `accounts-daemon`?

AWS EC2 instances have `accounts-daemon` :thinking:

j-rivero commented:
>
> Hmm, do you know if dokken supports linux native systemd units such as `accountsservice` which provides `accounts-daemon`?
> 
> AWS EC2 instances have `accounts-daemon` 🤔

dokken is no more than a docker with some helpers so it can support the systemd units.  Some points:

 * Is really needed the accounts-daemon ? What functionality is providing that was not used before?
 * The accounts-daemon can be installed in CI for dokken, it is not there by default
 * Even if the service exists, I think that Chef needs to know about it by declaring it a service in the chef code somewhere https://docs.chef.io/resources/service/#examples

Crola1702 committed:
Use noble for node names

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Declare accounts-daemon service

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Add citest url to kitchen config

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Configure nvidia only if has nvidia support

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Refactor has_nvidia_support conditions

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Add citest url to kitchen config

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Fix syntax errors

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Fix yamllint

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 commented:
Now CI is working @j-rivero.

I declared an `accounts-daemon` service. Refactored the code for nvidia support.

Most importantly, I want your check on 202f14b and 9607f77, as I added the citest jenkins URL

j-rivero reviewed (commented):
> Most importantly, I want your check on [202f14b](https://github.com/osrf/osrf_jenkins_agent/commit/202f14bb5e67f119f24be33b7c154ef296d3bf0b) and [9607f77](https://github.com/osrf/osrf_jenkins_agent/commit/9607f7774b13c27383d02146beffb4e5fcc7b086), as I added the citest jenkins URL

Ouk, why is this needed? I would not expect to see that value in a test configuration of this repo.

Crola1702 commented:
> Ouk, why is this needed? I would not expect to see that value in a test configuration of this repo.

So the definition of the jenkins url used in CI is:
https://github.com/osrf/osrf_jenkins_agent/blob/c9e6a3a206ac56d0a7b1ee0c4b3e8022a4aa7a5f/attributes/default.rb#L15

And it is used to define the agent connection in:
https://github.com/osrf/osrf_jenkins_agent/blob/c9e6a3a206ac56d0a7b1ee0c4b3e8022a4aa7a5f/recipes/default.rb#L73-L78


As this is not a real host, the test `Service jenkins-agent is expected to be running` fails (as the agent doesn't have anything to connect to). The alternative I propose is to add citest as the default jenkins URL for CI.

j-rivero commented:
> As this is not a real host, the test `Service jenkins-agent is expected to be running` fails (as the agent doesn't have anything to connect to). The alternative I propose is to add citest as the default jenkins URL for CI.

I think that mixing CI and testing environments is not a good idea. I've created https://github.com/osrf/osrf_jenkins_agent/pull/50 to resolve the problem by relaxing a test that can not pass without a fake server or similar.

nuclearsandwich commented on recipes/agent_build_tools.rb:
```diff
@@ -31,133 +30,112 @@
+  # Breaking X loading on AWS
+  package 'pulseaudio-module-bluetooth' do
+    action :purge
+  end
```
It might be worth a brief test to see if this quirk can be removed for 24.04.

j-rivero replied:
done in [a7d709b](https://github.com/osrf/osrf_jenkins_agent/pull/48/commits/a7d709b71656d68a848bac89e6f2abe94aff7df9).

nuclearsandwich commented on recipes/default.rb:
```diff
@@ -23,13 +23,13 @@
 include_recipe 'osrf_jenkins_agent::agent_build_tools' if node['osrfbuild']['agent']['install_agent_build_setup']
 
 # TODO: how to read attributes from chef-osrf plugins into this cookbook
 # swarm_client_version = node['jenkins-plugins']['swarm']
-swarm_client_version = "3.24"
+swarm_client_version = "3.49"
```
Hard coding this was surely an expedient on my part. Rather than doing a hard code, we could use the same attribute path and then this will "just work" when both the osrfbuild and jenkins agent cookbook are used in the same environment and otherwise the attribute in this cookbook can provide a sensible default.

nuclearsandwich commented on recipes/default.rb:
```diff
@@ -51,10 +51,10 @@
 
 if has_nvidia_support?
-  node_name = "linux-#{node_base_name}.nv.focal"
+  node_name = "linux-#{node_base_name}.nv.noble"
```
The codename is probably available in ohai somewhere (e.g. `node["lsb"]["codename"]` and we could make this a template rather than a hard coded value.

j-rivero replied:
done in [1db731f](https://github.com/osrf/osrf_jenkins_agent/pull/48/commits/1db731f59ac15062947bb659f22e626be9c67a42)

Commit:
CI: Aovid to start jenkins server in the test environment

* Remove citest

Signed-off-by: Jose Luis Rivero <jrivero@honurobotics.com>

* Service can not be running

Signed-off-by: Jose Luis Rivero <jrivero@honurobotics.com>

* Do not make assumptions on running service

* We can not start the service in testing. It is failing some of the times

Signed-off-by: Jose Luis Rivero <jrivero@honurobotics.com>

---------

Signed-off-by: Jose Luis Rivero <jrivero@honurobotics.com>

j-rivero committed:
Use lsb codename instead of hardcoding noble

Signed-off-by: Jose Luis Rivero <jrivero@honurobotics.com>

j-rivero committed:
Remove a previous problem on bluetooth

Signed-off-by: Jose Luis Rivero <jrivero@honurobotics.com>

j-rivero reviewed (approved):
I think that is good to go.
